lib/api/model/model.dart

```diff
@@ -305,6 +307,121 @@ sealed class Message {
   Map<String, dynamic> toJson();
 }
 
+/// A message's reactions, in a convenient data structure.
+class Reactions {
+  final List<Reaction> unaggregated;
+
+  /// A list of [ReactionWithVotes] objects.
+  ///
+  /// There won't be two items with the same
+  /// [ReactionWithVotes.reactionType] and [ReactionWithVotes.emojiCode].
+  /// (We don't also key on [ReactionWithVotes.emojiName];
+  /// see [ReactionWithVotes].)
+  late final List<ReactionWithVotes> aggregated;
+
+  Reactions(this.unaggregated) {
+    final byReaction = <int, ReactionWithVotes>{};
+    for (final reaction in unaggregated) {
```

Comment:
As far as efficiency is concerned, I think probably the most important single thing to do on this model is to minimize the work that's done for a message that has no reactions. That's probably the vast majority of messages.

A basic step to take in that direction would be to just give this constructor a short-cut path, where if the given list of reactions is empty then we skip making a map and just say the aggregated list is empty.

A more complete step would be to make `Message.reactions` nullable, with null representing empty, in the same way as `User.profileData`. That will make consuming code slightly more complicated; but I think the number of places in the code that will consume a message's reactions is fairly small, so that probably isn't too bad.